### ISSUE TITLE
create-function removed in PHP 8

### ISSUE THREAD

**Issue by arlt** [dokufreaks/plugin-pagelist#132]
Tested with 2022-07-31a "Igor" and:
```
$ cat /etc/centos-release
CentOS Linux release 7.9.2009 (Core)
$ php --version
PHP 8.1.10 (cli) (built: Aug 30 2022 16:09:36) (NTS gcc x86_64)
Copyright (c) The PHP Group
Zend Engine v4.1.10, Copyright (c) Zend Technologies
```

`create_function` ist removed in PHP 8 regarding to https://www.php.net/manual/en/function.create-function.php.
Used in: https://github.com/dokufreaks/plugin-pagelist/blob/master/syntax.php#L84

Since PHP 7 reaches EOL in 2022/11 and Dokuwiki Igor supports PHP 8, pagelist should support PHP 8, too.

```
<pagelist&table&header&nouser&nodate&tags>
  * [[wiki::syntax]]
  * [[wiki::welcome]]
</pagelist>
```
->
```
Error: Call to undefined function create_function()

An unforeseen error has occured. This is most likely a bug somewhere. It might be a problem in the pagelist plugin.

More info has been written to the DokuWiki error log.
```
->
```
Error: Call to undefined function create_function() /var/www/dokuwiki/lib/plugins/pagelist/syntax.php(84)
    #0 /var/www/dokuwiki/inc/parser/renderer.php(117): syntax_plugin_pagelist->render()
    #1 /var/www/dokuwiki/inc/parserutils.php(682): Doku_Renderer->plugin()
    #2 /var/www/dokuwiki/inc/parserutils.php(149): p_render()
    #3 /var/www/dokuwiki/inc/parserutils.php(89): p_cached_output()
    #4 /var/www/dokuwiki/inc/Ui/PageView.php(68): p_wiki_xhtml()
    #5 /var/www/dokuwiki/inc/Action/Show.php(37): dokuwiki\Ui\PageView->show()
    #6 /var/www/dokuwiki/inc/template.php(100): dokuwiki\Action\Show->tplContent()
    #7 [internal function]: tpl_content_core()
    #8 /var/www/dokuwiki/inc/Extension/Event.php(133): call_user_func_array()
    #9 /var/www/dokuwiki/inc/Extension/Event.php(199): dokuwiki\Extension\Event->trigger()
    #10 /var/www/dokuwiki/inc/template.php(85): dokuwiki\Extension\Event::createAndTrigger()
    #11 /var/www/dokuwiki/lib/tpl/dokuwiki/main.php(59): tpl_content()
    #12 /var/www/dokuwiki/inc/actions.php(27): include('...')
    #13 /var/www/dokuwiki/doku.php(126): act_dispatch()
    #14 {main} 
```

Additional warnings:
```
[Tue Sep 06 08:22:51.403223 2022] [proxy_fcgi:error] [pid 54210] [client 127.0.0.1:40704] AH01071: Got error 'PHP message: PHP Warning:  Undefined array key "page" in /var/www/dokuwiki/lib/plugins/pagelist/helper.php on line 220PHP message: PHP Warning:  Undefined array key "tags" in /var/www/dokuwiki/lib/plugins/pagelist/helper.php on line 226', referer: https://localhost/start?do=edit
```

**Comment by fiwswe** [comment]
Strange, I see the code containing `create_function(…)` [here](https://github.com/dokufreaks/plugin-pagelist/blob/master/syntax.php#L84) and I agree that [php.net](https://www.php.net/manual/en/function.create-function) documents this as deprecated in PHP 7.2.0 and removed in PHP 8.0.0. But I can't reproduce the error for some reason.

DokuWiki 2022-07-31 "Igor", PHP 8.0.23 (PHP-FPM), OpenBSD 7.1 amd64 running [httpd(8)](https://man.openbsd.org/OpenBSD-7.1/httpd), Pagelist 2022-06-26

Testpage:
```
====== Test the Pagelist Plugin ======

~~NOCACHE~~

[[doku>plugin:pagelist|PageList Plugin]]

<pagelist&table&header&nouser&nodate&tags>
  * [[wiki::syntax]]
  * [[wiki::welcome]]
</pagelist>
```

I am calling this page with `&purge` in addition to the `~~NOCACHE~~` to make sure there is no caching going on. All I see is a few PHP Warnings about unknown array indexes (see #131 and @arit mentioned them as well.)

Looking at the Pagelist code on my server, which is up to date according the Extension Manager, I see that it is completely different to what I see here in [master](https://github.com/dokufreaks/plugin-pagelist/blob/master/syntax.php). And http://www.dokuwiki.org/plugin:pagelist lists version 2020-09-27, not 2022-06-26.

Apart from my failure to reproduce, the recommended replacement seems to be an [anonymous function](https://www.php.net/manual/en/functions.anonymous.php). That is also what the version I have on my server uses.

So why is [master](https://github.com/dokufreaks/plugin-pagelist/blob/master/syntax.php) showing apparently old code?

Is maybe [cleanup](https://github.com/dokufreaks/plugin-pagelist/blob/cleanup/syntax.php) the branch the published plugin is based on? It seems to match what I have on my server.

fiwswe

**Comment by splitbrain** [comment]
cleanup is what is in #130 - maybe provide feedback for @Klap-in there so that it can be merged.

**Comment by fiwswe** [comment]
Ah, now I see! I followed @Klap-in's [suggestion](https://github.com/dokufreaks/plugin-pagelist/issues/131#issuecomment-1229431977) to try the cleanup version to help fix #131. That is why I have the newer version installed.

It does fix some, but not all PHP Warnings and also solves #132.

I still see:
```
PHP message: PHP Warning:  Undefined array key "type" in /dokuwiki-test/lib/plugins/pagelist/helper.php on line 662
PHP message: PHP Warning:  Undefined array key "type" in /dokuwiki-test/lib/plugins/pagelist/helper.php on line 662
```
on the above mentioned test page.

Otherwise a merge of #130 would certainly be an improvement.


fiwswe.

**Comment by arlt** [comment]
@Klap-in `cleanup.zip` from #131 solved this issue here for me, too.
Would be nice if it can be merged and released.

Remaining warnings:

```
[Tue Sep 06 12:27:19.451535 2022] [proxy_fcgi:error] [pid 54210] [client 127.0.0.1:43970] AH01071: Got error 'PHP message: PHP Warning:  Undefined array key "type" in 
/var/www/dokuwiki/lib/plugins/pagelist/helper.php on line 662PHP message: PHP Warning:  Undefined array key "type" in /var/www/dokuwiki/lib/plugins/pagelist/helper.php on line 662', referer: https://localhost/start?do=edit
```

These errors occured, too - but i am not sure because they don't appear reliable:
```
[Tue Sep 06 12:27:19.779275 2022] [proxy_fcgi:error] [pid 54210] [client 127.0.0.1:43970] AH01071: Got error 'PHP message: PHP Warning:  Undefined array key 1 in 
/var/www/dokuwiki/lib/plugins/pagelist/syntax.php on line 59PHP message: PHP Warning:  Undefined array key 2 in 
/var/www/dokuwiki/lib/plugins/pagelist/syntax.php on line 59PHP message: PHP Warning:  Undefined array key 1 in 
/var/www/dokuwiki/lib/plugins/pagelist/syntax.php on line 60PHP message: PHP Warning:  Undefined array key 1 in 
/var/www/dokuwiki/lib/plugins/pagelist/syntax.php on line 59PHP message: PHP Warning:  Undefined array key 2 in 
/var/www/dokuwiki/lib/plugins/pagelist/syntax.php on line 59PHP message: PHP Warning:  Undefined array key 1 in 
/var/www/dokuwiki/lib/plugins/pagelist/syntax.php on line 60', referer: https://localhost/start
```

**Comment by fiwswe** [comment]
> These warnings occured, too - but i am not sure because they don't appear reliable:

[list()](https://www.php.net/manual/de/function.list) statements with unverified inputs. The root cause is probably wherever the exploded strings come from.

But this shows the dangers of using `list()`once again! `list()` is great if you are 100% sure that the source array contains the requested elements. Otherwise it should not be used IMHO.

fiwswe

**Comment by fiwswe** [comment]
```
            list($id, $title, $description) = explode('|', $match[1], 3);
            list($id, $section) = explode('#', $id, 2);
```
could be replaced by the ugly but more correct:
```
            $tmp = explode('|', $match[1], 3);
            // $id = $tmp[0]; // Pointless because it is set again a few lines down.
            if (count($tmp) > 1)
                $title = $tmp[1];
            else
                $title = null;
            if (count($tmp) > 2)
                $description = $tmp[2];
            else
                $description = null;
            $tmp = explode('#', $id, 2);
            $id = $tmp[0];
            if (count($tmp) > 1)
                $section = $tmp[1];
            else
                $section = null;
```
Do we need to initialize `$title`, `$description` and `$section` to some different default value if the source string does not contain the fields? I haven't really taken the time to understand the rest of the code.

fiwswe

**Comment by splitbrain** [comment]
see https://github.com/splitbrain/dokuwiki/pull/3754 for how to fix explode/list constructs. Please also provide pull requests instead of code comments to make it easier for @Klap-in

**Comment by fiwswe** [comment]
PR in #133 (against cleanup)

fiwswe

**Comment by Klap-in** [comment]
Fixed since merge of #130
